WatchFaceAlphaKotlin/app/src/main/java/com/example/android/wearable/alpha/editor/WatchFaceConfigViewModel.kt

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2021 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.alpha.editor
+
+import android.content.Intent
+import android.graphics.Bitmap
+import android.graphics.Color
+import android.util.Log
+import androidx.activity.ComponentActivity
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import androidx.wear.complications.data.ComplicationData
+import androidx.wear.watchface.DrawMode
+import androidx.wear.watchface.RenderParameters
+import androidx.wear.watchface.editor.EditorSession
+import androidx.wear.watchface.style.UserStyle
+import androidx.wear.watchface.style.UserStyleSetting
+import androidx.wear.watchface.style.WatchFaceLayer
+import com.example.android.wearable.alpha.data.watchface.MINUTE_HAND_LENGTH_FRACTION
+import com.example.android.wearable.alpha.data.watchface.MINUTE_HAND_LENGTH_FRACTION_MAXIMUM
+import com.example.android.wearable.alpha.data.watchface.MINUTE_HAND_LENGTH_FRACTION_MINIMUM
+import com.example.android.wearable.alpha.utils.COLOR_STYLE_SETTING
+import com.example.android.wearable.alpha.utils.DRAW_HOUR_PIPS_STYLE_SETTING
+import com.example.android.wearable.alpha.utils.WATCH_HAND_LENGTH_STYLE_SETTING
+import java.beans.PropertyChangeListener
+import java.beans.PropertyChangeSupport
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+
+/**
+ * Handles all ops for saving changes from [WatchFaceConfigActivity] to the watch face data layer,
+ * that is to the [EditorSession].
+ *
+ * Uses the keys, [UserStyleSetting], for each of our user styles and sets their values
+ * [UserStyleSetting.Option]. After a new value is set, creates a new image preview via screenshot
+ * class and triggers a listener (which creates new data
+ * for the [StateFlow] that feeds back to the Activity).
+ */
+class WatchFaceConfigViewModel(activity: ComponentActivity, editIntent: Intent) : ViewModel() {
+    private val changeSupport = PropertyChangeSupport(this)
```

Comment:
The `PropertyChangeSupport` looks really suspicious. We should be able to use a `kotlinx.coroutines` primitive versus a callback/listener setup, and ideally even be able to replace the imperative "notify about an update" into a pure state flow.

Reply:
Will revise and follow up.

Reply:
Done; Removed callbackFlow and used a regular StateFlow.